sdk/containerregistry/container-registry/src/containerRegistryClient.ts

```diff
@@ -110,4 +119,86 @@ export class ContainerRegistryClient {
       span.end();
     }
   }
+
+  /**
+   * Iterates repositories.
+   *
+   * Example usage:
+   * ```ts
+   * let client = new ContaienrRegistryClient(url, credentials);
+   * for await (const repository of client.listRepositories()) {
+   *   console.log("repository name: ", repository);
+   * }
+   * ```
+   * @param options -
+   */
+  public listRepositories(
+    options: ListRepositoriesOptions = {}
+  ): PagedAsyncIterableIterator<string, string[]> {
+    const { span, updatedOptions } = createSpan("listRepositories", options);
+    const iter = this.listRepositoryItems(updatedOptions);
+
+    span.end();
+    return {
+      next() {
+        return iter.next();
+      },
+      [Symbol.asyncIterator]() {
+        return this;
+      },
+      byPage: (settings: PageSettings = {}) => this.listRepositoriesPage(settings, updatedOptions)
+    };
+  }
+
+  private async *listRepositoryItems(
+    options: ListRepositoriesOptions = {}
+  ): AsyncIterableIterator<string> {
+    for await (const page of this.listRepositoriesPage({}, options)) {
+      yield* page;
+    }
+  }
+
+  private async *listRepositoriesPage(
+    continuationState: PageSettings,
+    options: ListRepositoriesOptions = {}
+  ): AsyncIterableIterator<string[]> {
+    if (!continuationState.continuationToken) {
+      const optionsComplete = {
+        n: continuationState.maxPageSize,
+        ...options
+      };
+      const currentPage = await this.client.containerRegistry.getRepositories(optionsComplete);
+      // TODO: (jeremymeng) nextLink not working as expected
+      // continuationState.continuationToken = currentPage.link;
+      if (currentPage.names) {
+        continuationState.continuationToken = currentPage.names[currentPage.names.length - 1];
+        yield currentPage.names;
+      } else {
+        continuationState.continuationToken = undefined;
+      }
+      while (continuationState.continuationToken) {
+        // TODO: (jeremymeng) nextLink not working as expected
+        // const currentPage = await this.client.containerRegistry.getRepositoriesNext(
```

Comment:
The next link value returned by the service is not in the expect format.  It is of the form `</acr/v1/library%2Falpine/_manifests?last=sha256%3A43abbf92155ae0e7067e9b619bd70d4e14f6bf6adef026d336f5aed3d4b5d6a7&n=2&orderby=>; rel="next"`, while we are expecting just the path part `/acr/v1/library%2Falpine/_manifests?last=sha256%3A43abbf92155ae0e7067e9b619bd70d4e14f6bf6adef026d336f5aed3d4b5d6a7&n=2&orderby=`.  So we have to do custom paging in convenience layer, either parsing the path from next link value, or use the last item in the current page's result as next marker.

Reply:
@jeremymeng, are there any transformations we can do in the swagger file or autorest.md to parse this value for the code generator?

Reply:
Technically it is possible, as we could do transformation on generated code files (for example, [the original readme](https://github.com/Azure/azure-rest-api-specs/tree/255757f41275e8ec474361690ea8886cae8a503b/specification/containerregistry/data-plane#tag-package-2019-08) has this for generated csharp):
```
directive:
  - from: source-file-csharp
    where: $
    transform: >-
      return $.
        replace( /_url = _url.Replace\("\{nextBlobUuidLink\}", location\);/g, "_url = _url.Replace(location.StartsWith(\"/\") ? \"/{nextBlobUuidLink}\" : \"{nextBlobUuidLink}\", location);")
```
For this case I was concerned that the code we want to replace is too generic. And the transformation will be different for each language. So I haven't gone down this path yet.

Reply:
Do we know for sure there's not something we could do in the shared readme.md that would be applicable across all languages?

Reply:
I've been discussing this with @joheredi but so far we don't have a good solution.

Reply:
Unfortunately transforms like the one Jeremy pointed out above are language specific and are basically matching a string in the generated code and replacing by another string.

The sample above is looking for a specific line of code generated by the C# generator and replacing it with a different one. This is very specific to C# and none of the other languages would be able to reuse. In the best case scenario each language would need a specific transform like this.

Unfortunately, the service seems to diverge from the `x-ms-pageable` extension standard, in which nextLink would be in the format ready to be used. Do we know the reason why the service is returning the nextLink in the format they do and not in the format that is needed to get the next page?

Usually when services don't implement features like Pagination according to the standard, the best approach is to remove the `x-ms-pageable` extension and handle pagination completely in the convenience layer. The reason is that the generators are designed to understand the standard extensions, forcing a non-standard swagger to fit into it may cause more troubles than it is worth in the future

Reply:
> Do we know the reason why the service is returning the nextLink in the format they do and not in the format that is needed to get the next page?

I believe it is because the v1 endpoint is following Docker API spec.  The service might consider adding `next` in response body that can be used as a continuation token.

> the best approach is to remove the `x-ms-pageable` extension and handle pagination completely in the convenience layer
It is still useful in that we can extract the part we need to pass to getXxxxNext() method. IMO it's better than passing last result as `last` and `n` to the getXxxx() method.